pkg/xds/experimental/client/watcher.go

```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package client
+
+import (
+	"log/slog"
+
+	"github.com/cilium/cilium/pkg/envoy/xds"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// watchers is a helper structure for Client focused on [un]registering callbacks on resource changes.
+type watchers struct {
+	log *slog.Logger
+
+	src xds.ObservableResourceSource
+
+	// mux protects all fields below.
+	mux lock.Mutex
+	// listeners maps ID to an instance of a listener.
+	listeners map[uint64]*listener
+	lastID    uint64
+}
+
+func newWatchers(log *slog.Logger, src xds.ObservableResourceSource) *watchers {
+	return &watchers{
+		log:       log,
+		src:       src,
+		listeners: make(map[uint64]*listener),
+	}
+}
+
+// WatcherCallback will be called when a new version of a resource it was registered on appears.
+type WatcherCallback func(res *xds.VersionedResources)
+
+// Add registers a callback for a specified typeUrl.
+// Returned ID can be used to later unregister the callback.
+func (w *watchers) Add(typeUrl string, cb WatcherCallback) uint64 {
+	w.mux.Lock()
+	defer w.mux.Unlock()
+
+	w.lastID++
+
+	l := &listener{
+		ID:        w.lastID,
+		typeUrl:   typeUrl,
+		cb:        cb,
+		log:       w.log.With("listenerID", w.lastID),
+		resources: w.src,
+		trigger:   make(chan uint64, 1),
+	}
+	go l.process()
+	w.src.AddResourceVersionObserver(l)
```

Comment:
Is this function blocking or long? In comment src is excluded from mutex comment but still you lock whole function Add and Remove. Is this needed?

Reply:
I don't know what "long" function is. Could please tell me if [this one](https://github.com/cilium/cilium/blob/6a75f3d2087a55fe2c757bc883630223e6841370/pkg/envoy/xds/set.go#L147-L154) is?

Reply:
xds.ObservableResourceSource is an interface and it can be implemented in different ways if not now then in future for sure it will change. You should explicitly says that this operation is called behind the lock and should not be blocking

Reply:
This function is inherently fast (as the name suggests it justs adds an observer) and I don't foresee it having any other implementation (it is not an API surface, it will be used only within Cilium repo)